Feature/1468 sweepformula epsp

t-b opened the pull request.

t-b commented:
@timjarsky We can chat abour this on monday.

t-b commented:
@timjarsky @travis-open So here is the second final version.

Although we are pretty close to finalize things, please don't create production data with this version as it still might change in an incompatible way.

Some SF example code to start playing around.

```
# Preliminary documentation
# 
# epsp([epspKernel(), peakThresh, filterLow, filterHigh])
#
# epspKernel([select(), tau1, tau2, amp])
# 
# epspStats(prop, select(), state, [postProc])
#   prop:     amp/x-pos/x-interval/tau
#   state:    accepted/rejected/all
#   postProc: nothing/avg/count/hist

# Examples:

# categorize events from the current sweep  with default values for the parameters
epsp()

# From sweep 50 / AD10
# epsp(epspKernel(select(channels(AD10), 50, all)))

# same version but passing all default parameters
# epsp(epspKernel(select(channels(AD10), 50, all), 1, 15, (-5)), 0.01, 550, 0)

and

# display the default epsp kernel waves and sweep data
# epspKernel()

and

# display some stats

epspStats(tau, select(channels(AD10), 50, all), all)

and

epspStats("x-pos", select(channels(AD10), 50, all), rejected, hist)


and

epspStats(amp, select(channels(AD10), 50, all), accepted, count)
```

timjarsky commented:
@t-b `epspKernel` SF should accept `data` and `select` as users will want to analyze sub-regions of their sweeps.

timjarsky commented:
@t-b sweeps that are terminated early will have NaNs. 
```
  !!! Assertion FAILED !!!
  Message: "Can not handle NaN in the deconvoluted wave"
  Please provide the following information if you contact the MIES developers:
  ################################
```

timjarsky commented:
@t-b when working with the linked file I get:

`epsp(epspKernel(select(channels(AD9), 30, all), 1, 15, (-5)), 0.01, 550, 0)` or `epsp(epspKernel(select(channels(AD9), 30, all)))`

```
  ################################
  !!! Assertion FAILED !!!
  Message: "Missing datafolder"
  Please provide the following information if you contact the MIES developers:
  ################################
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Stacktrace:
SF_button_sweepFormula_display(...)#L2233 [MIES_SweepFormula.ipf]
SF_FormulaPlotter(...)#L1505 [MIES_SweepFormula.ipf]
SF_GatherFormulaResults(...)#L1036 [MIES_SweepFormula.ipf]
SF_ExecuteFormula(...)#L4535 [MIES_SweepFormula.ipf]
SF_FormulaExecutor(...)#L1007 [MIES_SweepFormula.ipf]
ES_Operation(...)#L483 [MIES_SweepFormula_EPSP.ipf]
ES_RestoreEventState(...)#L430 [MIES_SweepFormula_EPSP.ipf]
ES_GetEventsFromResults(...)#L1138 [MIES_SweepFormula_EPSP.ipf]
BSP_GetLogbookWave(...)#L1423 [MIES_BrowserSettingsPanel.ipf]
SB_GetLogbookWave(...)#L527 [MIES_AnalysisBrowser_SweepBrowser.ipf]
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Time: 2023-01-13T11:55:43-08:00
  Locked device: [- none -]
  Current sweep: [- none -]
  DAQ: [- none -]
  Testpulse: [- none -]
  Experiment: Untitled ()
  Igor Pro version: 9.0.1.2 (39200)
  MIES version:
  Release_2.6_20220606-746-g67c5ca84
Date and time of last commit: 2023-01-13T17:38:31+01:00
Submodule status: 
-e46385bee5c53a0daa1eca0c1d6af44d19677c13 Packages/IPNWB
-d229d451f35e88b8c38edb8c26533acfbc3fd10f Packages/doc/doxygen-filter-ipf
-873dee901d747ceedfce01e66c058c8791a3b046 Packages/unit-testing
  ################################
```
[2023_01_10_152030-compressed.zip](https://github.com/AllenInstitute/MIES/files/10414955/2023_01_10_152030-compressed.zip)

t-b commented:
- [ ] epspKernel SF should accept data and select as users will want to analyze sub-regions of their sweeps.

Good point. I'll think about something.

- In addition to subregions, users will also want to analyze multiple sweeps at once. Why is it limited to one sweep?

Because the UI will get really ugly with more than one sweep. For example the single event view will be duplicated and it will be difficult to connect it to the event plot with the cursor. And it will get even more crowded once we plot the overlayed events.

- [ ] sweeps that are terminated early will have NaNs.

That is probably an older sweep without unacquired epoch. I'll have a look.

- [ ] when working with the linked file I get

Nice find!

timjarsky commented:
> * In addition to subregions, users will also want to analyze multiple sweeps at once. Why is it limited to one sweep?
> 
> Because the UI will get really ugly with more than one sweep. For example the single event view will be duplicated and it will be difficult to connect it to the event plot with the cursor. And it will get even more crowded once we plot the overlayed events.
> 


@t-b Instead of duplicating, is it possible to treat multiple sweeps or epochs as one long sweep concatenated together?

t-b commented:
From chat with TJ:

- [ ] epspKernel should accept data operation instead of select
- [ ] Conclusion regarding multiple sweeps for epsps. We do that as this is the natural way to think about the data. TJ will ask TH if we want to have relative time or absolute time. Needs to be able to figure out the original event time in any case.

timjarsky commented:
@t-b 
> * [ ]  epspKernel should accept data operation instead of select

Yes, but running the kernel independently on each data chunk is okay. This should work well with data acquisition as we could avoid rerunning the kernel on previously acquired data. Let's discuss parallelizing this step to speed it up. 

Are steps taken to optimize the length of the data for the FFT operation?

> * [ ]  Conclusion regarding multiple sweeps for epsps. We do that as this is the natural way to think about the data. TJ will ask TH if we want to have relative time or absolute time. Needs to be able to figure out the original event time in any case.

The natural way to think about the data is as sweeps/epochs, so let's go with a list box that allows the user to select which sweep to QC. The sweep list is determined by data operation.

timjarsky commented:
@t-b 

Here is an example SF I want to verify we will support:

epsp(epspKernel(data(E0, select(channels(AD10), [1...10], all)))

and

epsp(epspKernel(data(E0, select(channels(AD10), [10...20], all)))

and

epspStats(amp, data(E0, select(channels(AD10), [1...10], all), accepted, hist)

with

epspStats(amp, data(E0, select(channels(AD10), [10...20], all), accepted, hist)

t-b commented:
@timjarsky 

> Here is an example SF I want to verify we will support:

From my current understanding I would say that both epsp operations can be combined into 
`epsp(epspKernel(data(E0, select(channels(AD10), [1...20], all)))` or? My reasoning is that we:
- want to have one epsp plot
- where the user selects the sweeps from the listbox
- And do the kernel per sweep/epoch

> epspStats(amp, data(E0, select(channels(AD10), [1...10], all), accepted, hist)

After we move `epspStats` to accept data, this will work. Do you expect to have per sweep traces are one trace?

timjarsky commented:
@t-b We do want to support both examples. In my example, the kernel is different (sorry, I shouldn't have left the default values) because an experimental treatment changed the properties of the synaptic response.

Let's talk about the best or easiest way to handle two kernels at once.

t-b commented:
- Store the epsp data in a datafolder per sweep/channel/range, same layout with results wave
- epspStats gains a new optional parameter at the end `comb` (default) /`sep` which says if the event data from multiple sweeps should be combined or kept separate.

t-b commented:
@timjarsky Ready for another shot, documentation is at https://github.com/AllenInstitute/MIES/pull/1468#issuecomment-1382112810.

timjarsky commented:
Mini QC topics to demon in next chat:
1) keyboard shortcuts after scaling - must select the main panel. selecting sub panel does not work
2) auto move to the last qc'd event
3) histogram of tau values has strange values
4) mini categories; all, accepted, rejected, reviewed 
5) Constrain fitting - create ways to accept event but reject fit, for example
6) additional operations to return histogram features
7) constraining curve fit - direction? of exponential
8) show the last, current, and next event in zoom view. Have the past and upcoming in smaller boxes
9) tau event number doesn't match between tau stats display and epsp operation plot
10) placing other cursors on the SF plots results in errors
11) event start time often preceeds event 
12) last event in a sweep
```
  BUG_TS: Encountered pending RTE: 1321, a wave read;Index out of range for wave "epspEvent".
  ** a wave read gave error: Index out of range for wave "epspEvent".
  ** a wave read gave error: Index out of range for wave "epspEvent".
  ** a wave write gave error: Index out of range for wave "epspEvent".
```
13) consider removing dashes from operation parameter names so quotes aren't necessary

t-b commented:
@timjarsky Not all todos resolved, still I think this is worth to give a shot again. Now works with multiple range/select combinations and multiple epsp operation calls.

timjarsky commented:
> @timjarsky Not all todos resolved, still I think this is worth to give a shot again. Now works with multiple range/select combinations and multiple epsp operation calls.

@t-b can you share an example formula? I tried substituting the data operation for the select operation but I get a range error.

t-b commented:
@timjarsky The example above, see https://github.com/AllenInstitute/MIES/pull/1468#issuecomment-1382112810, has some code. I'm using range + select instead of data.

timjarsky commented:
@t-b 

1) make the jump button work across sweeps/regions.
2) After storing results, `Could not find any EPSP events in the results wave error` with 
```epspStats([2000,4000], select(channels(AD9), 30, all), tau, accepted)```
3) Why use `range` in combination with `select` instead of `data` operaiton?

timjarsky commented:
@t-b I'd like to adjust the epsp plot legend to get it to fit above the plot. Something like:
![image](https://user-images.githubusercontent.com/5693697/220475724-b200652f-0de0-44d7-8511-b92a60acdda6.png)

t-b commented:
@timjarsky 

I thought about something like

![image](https://user-images.githubusercontent.com/1185677/220655344-3eeeb6f9-b33a-471d-ba02-9a0c818b82a6.png)

instead of packing the legend on top?

timjarsky commented:
@t-b An advantage of positioning the info box at the top is that the space is unused. It's too bad text boxes don't support columns/tables, which makes formatting, as in my example above, challenging. So, perhaps, the simplest thing is to have two text boxes at the top without a frame?

timjarsky commented:
@t-b wow, it is looking really nice. I have a minor feature request. I want to be able to plot two additional states. Fit failed, and fit passed.

timjarsky commented:
@ t-b, There may be issues related to accepting/rejecting events and fits when the fit fails. e.g., using a marquee to reject events, where the fit failed, gives: 

```
  BUG_TS: Encountered pending RTE: 193, FilterIIR;expected /HI frequency between 0 and 0.5```

t-b commented:
>   BUG_TS: Encountered pending RTE: 193, FilterIIR;expected /HI frequency between 0 and 0.5```

[]I'm not really working today, just passing by ;)] Are you using custom filter frequencies for `epsp`? This is the only time we use `FilterIIR` in all of EPSP. Do you get that error with the default values as well? Parameter order switched?

t-b commented:
@timjarsky 

> I have a minor feature request. I want to be able to plot two additional states. Fit failed, and fit passed.

I have added `fitresult` as new property to plot for `epspStats`.

timjarsky commented:
@t-b feature requests for individual and average event plot. (1) Add a combined range checkbox that induces other checkboxes to plot events and averages from all ranges. e.g., all accepted events from all ranges. (2) maintain axes scaling after the user has modified scaling

![image](https://user-images.githubusercontent.com/5693697/223180929-7962ece6-4b8a-44af-a818-47020e6919b2.png)

t-b commented:
@timjarsky 
> Add a combined range checkbox that induces other checkboxes to plot events and averages from all ranges. e.g., all accepted events from all ranges.

Do you want one trace for *all* ranges per state or one trace per range and state?

t-b commented:
@timjarsky I've fixed mouse selection and added a couple of tests for it. Rest follows not today...

timjarsky commented:
@t-b I think a fitstate is needed. e.g., fit_accept

The goal is to allow the user to display a histogram of all the accepted fits.

The current behavior * I think * is that the following formula displays the tau values from all accepted events independent of whether the fit is accepted.

```epspStats([2000,4000], select(channels(AD9), [31,33,34], all), tau, accept, hist)```

t-b commented:
> @t-b additional shortcut key request; single key for accept the event and fail fit.

Done, use `z` for that. See https://github.com/AllenInstitute/MIES/pull/1468#issuecomment-1382112810.

timjarsky commented:
> > The current behavior * I think * is that the following formula displays the tau values > from all accepted events independent of whether the fit is accepted.
> 
> > epspStats([2000,4000], select(channels(AD9), [31,33,34], all), tau, accept, hist)
> 
> Well add another parameter?
> 
> Something like:
> 
> ```
> # epspStats(array range, select(), prop, state, [stateType, postProc])
> #
> # range:     range parameter, see also `data()`
> # prop:      amp/xpos/xinterval/tau/estate/fstate/fitresult
> # state:     accept/reject/undetermined/all/every
> # stateType: event/fit (default "event")
> # postProc:  nothing/avg/count/hist/log10 (default "nothing")
> ```

@t-b Adding another parameter would provide the most flexibility, but I wonder if it is necessary. It might be "smarter" to have the `state` be associated with the `prop`. In this case, that would mean any fit parameter (we currently only have tau) would use the fit state. For event parameters (mp/xpos/xinterval) the state would automatically be from the event.

timjarsky commented:
@t-b yes, the associations you outlined make sense. Thanks for fixing them.

t-b commented:
> @t-b, did "go to event" using marquee get dropped from the to-do list? If yes, let's add it back.

This was added as issue in #1641. Should it be added here?

> I'd also like all the marquee options to work from the subwindow plot in addition to the main SF window. e.g., a user selects one or more traces and can pass/fail them.

Let's chat about that today.

t-b commented:
@timjarsky I've added a new label: https://github.com/AllenInstitute/MIES/labels/Miniature%20Analysis.

timjarsky commented:
@t-b I don't believe this :arrow_down:  is working. I can demo if you have time.

```
 State should be associated with the prop:
amp/xpos/xinterval/estate -> event state
tau/fstate/fitresult -> fit state
```

t-b commented:
> @t-b I don't believe this ⬇️ is working. I can demo if you have time.
> 
> ```
>  State should be associated with the prop:
> amp/xpos/xinterval/estate -> event state
> tau/fstate/fitresult -> fit state
> ```

I don't think your version has that. I've just checked it as my local version has it. But I still want to get the all event graph working before I push a new version.

timjarsky commented:
@t-b Thanks for letting me know. Your plan sounds good to me.

t-b commented:
@timjarsky I've pushed an update. I think the all event graphs (top left) is now working as it should. I'm also beaten so I don't think I can fix the remaining issues today.

- For the all events plot only left mouse button changes are left to do
- You can now move the A cursor across combos
- Cursor A in epspStats plot now moves cursor A in epsp plot *and* all keyboard shortcuts work as well. I figured while at it I can just make that happen.

timjarsky commented:
@t-b I'm getting an assertion with

```epspStats([E0,E2], select(channels(AD0), 4...63, all), tau, all)```

```
  ################################
  !!! Assertion FAILED !!!
  Message: "Expected only a single epoch in the textual range"
  Please provide the following information if you contact the MIES developers:
  ################################
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Stacktrace:
SF_button_sweepFormula_display(...)#L2286 [MIES_SweepFormula.ipf]
SF_FormulaPlotter(...)#L1523 [MIES_SweepFormula.ipf]
SF_GatherFormulaResults(...)#L1036 [MIES_SweepFormula.ipf]
SF_ExecuteFormula(...)#L4565 [MIES_SweepFormula.ipf]
SF_FormulaExecutor(...)#L1013 [MIES_SweepFormula.ipf]
ES_OperationStats(...)#L3255 [MIES_SweepFormula_EPSP.ipf]
ES_OperationStatsImpl(...)#L854 [MIES_SweepFormula_EPSP.ipf]
ES_GenerateComboKey(...)#L2031 [MIES_SweepFormula_EPSP.ipf]
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Time: 2023-03-22T13:15:23-07:00
  Locked device: [- none -]
  Current sweep: [- none -]
  DAQ: [- none -]
  Testpulse: [- none -]
  Experiment: PSPtesting_artifactIssue (Packed)
  Igor Pro version: 9.0.1.2 (39200)
  MIES version:
  
  ################################
```

t-b commented:
@timjarsky 

> @t-b I'm getting an assertion with
epspStats([E0,E2], select(channels(AD0), 4...63, all), tau, all)

Fixed it.

t-b commented:
Added accepted average fitting.

t-b commented:
The accepted average fit is now done with dblexp_peak which is more robust than dblexp_offset.

timjarsky commented:
@t-b, the fit should start at the event onset time, not time zero. 

![image](https://user-images.githubusercontent.com/5693697/234674258-a876d874-e46d-47b1-a09c-09d6575e13b9.png)

We should also consider ways of limiting the amount of post-event baseline used for the fit

timjarsky commented:
@t-b there is also a compile error resulting from an incorrectly named optional param for `isWaveLargeEnough`

t-b commented:
> @t-b, the fit should start at the event onset time, not time zero.

Use average of event start time, calculate also the average of the event lengths and use that as fit length.

t-b commented:
I've adapted the fitting range to something better. If it still fails, I know also output the range as debug output. We can debug that together later as well.

timjarsky commented:
@t-b Would it be possible to plot Xinterval against event no. by giving the last event a NaN interval? This would make the plotting of Xinterval useful, especially in the case of the histogram.

I'm also seeing an issue where `all' and 'every` fails to plot - I'll upload an example pxp to your FTP later today. In these cases, `accepted` works as expected.

If the `store` function is used with psxStats...hist, I believe it stores the counts in each bin but not the x value of the bin. What's the recommended way to access the x info?

timjarsky commented:
> I've adapted the fitting range to something better. If it still fails, I know also output the range as debug output. We can debug that together later as well.

@t-b Looks better - no errors. But the fit is shifted relative to the trace:
![image](https://github.com/AllenInstitute/MIES/assets/5693697/73cc4727-feda-43db-ae95-03562dfc6522)

Also, let's add an option to make the decay fit with a double exponential.

t-b commented:
Q: Would it be possible to plot Xinterval against event no. by giving the last event a NaN interval? This would make the plotting of Xinterval useful, especially in the case of the histogram.

-> Use x-axis of event numbers like 1.5/2.5/... for x-interval.

Q: I'm also seeing an issue where `all' and 'every` fails to plot - I'll upload an example pxp to your FTP later today. In these cases, `accepted` works as expected.

A: Will be commited with hacky patch.

Q: If the `store` function is used with psxStats...hist, I believe it stores the counts in each bin but not the x value of the bin. What's the recommended way to access the x info?

A: Change `store` format to also use JSON serialization which should take care of dimdelta/offset. Adapt our other internal users of `store`.

t-b commented:
Ready for another look.

timjarsky commented:
@t-b So, I attempted to analyze an experiment from scratch, and everything worked great until I added an additional PSXstats formula (last line in the block shown below) which caused all the QC to disappear - it's like I changed the kernel or the data selection, but I didn't. I uploaded the psx_ALLqcDisappeared.pxp to your FTP.  Other than this, everything worked really well.

```
epoch=[4000,6000]
sel=select(channels(AD9),[65,66],all)

psx(psxKernel($epoch,$sel, 0.6, 3, (-6)), 0.04, 550, 0)

and

psxStats($epoch,$sel, tau,acc)

and

psxStats($epoch,$sel, tau,acc, hist)

and

psxStats($epoch,$sel, amp,acc, hist)

and

psxStats($epoch,$sel, amp,all, hist)
```

t-b commented:
> Other than this, everything worked really well.

Thanks for the enthusiastic reply which basically says "The software ate my data but it fast!". I'll have a look.

t-b commented:
@timjarsky I'm looking into psx_ALLqcDisappeared.pxp. Can you upload the file 2022_06_01_091630-compressed.nwb with the original data?

timjarsky commented:
> @timjarsky I'm looking into psx_ALLqcDisappeared.pxp. Can you upload the file 2022_06_01_091630-compressed.nwb with the original data?

I've put the file on the FTP @t-b

t-b commented:
@timjarsky > So, I attempted to analyze an experiment from scratch, and everything worked great until I added an additional PSXstats formula (last line in the block shown below) which caused all the QC to disappear.

So I reread what you said here and did you do the following:
- Write SF code and press display
- QC data
- Press display again
?

If yes that is how it currently is. IIRC we moved away for storing every change to each event at some point due to speed issues.

To solve the issue we can:
1. Automatically store when closing the SweepFormula plot with PSX QC'ed data. This has the disadvantage that we don't really know if the user really wants to save it.
2. Automatically keep unstored PSX QC data when pressing Display and refetch
3. Ask when Display finds PSX QC data if it should be kept
4. Keep as it is and document

My favourite would be 2.

t-b commented:
New syntax:

```
sel    = select(channels(AD10), [48, 49, 50], all)
sel49  = select(channels(AD10), [49], all)
xrange = [300, 500]
id     = someName

# psx(id, [psxKernel(), peakThresh, filterLow, filterHigh])
#
# psxKernel([range, select(), tau1, tau2, amp])
# 
# psxStats(id, array range, select(), prop, state, [stateType, postProc])
#
# range:     range parameter, see also `data()`
# prop:      amp/xpos/xinterval/tau/estate/fstate/fitresult
# state:     accept/reject/undetermined/all/every
# postProc:  nothing/avg/count/hist/log10 (default "nothing")

psx($id, psxKernel($xrange, $sel))

and

# display some stats

psxStats($id, $xrange, $sel, tau, all, h)

and

psxStats($id, $xrange, $sel49, tau, all, h)

and

psxStats($id, $xrange, $sel, amp, every)
```

t-b commented:
>  psxStats with xinterval, see file from Tim, select an event which is failing QC but we requested acc

I fixed a related bug already. But even then this is not too suprising as `xinterval` currently fiddles with the event index (+0.5) and then the selection gives wrong results. The most easiest fix would be to not fiddle with the event index.

timjarsky commented:
> > psxStats with xinterval, see file from Tim, select an event which is failing QC but we requested acc
> 
> I fixed a related bug already. But even then this is not too suprising as `xinterval` currently fiddles with the event index (+0.5) and then the selection gives wrong results. The most easiest fix would be to not fiddle with the event index.

@t-b I think that's fine. Would the first interval take on event index 1 because event indexing starts @ zero?

t-b commented:
>  Would the first interval take on event index 1 because event indexing starts @ zero?

Yes!

t-b commented:
Ready for another shot.

timjarsky commented:
@ t-b PSXstats xinterval is miscalculating a single interval very close to the start of a new range of data. This should be reproducible from files I previously shared but let me know if you need a new file.

![image](https://github.com/AllenInstitute/MIES/assets/5693697/deada9e9-26f3-4ad2-92af-09fc3486b669)

t-b commented:
Not sure why I'm still assigned. I'll pass it on to @timjarsky.

timjarsky commented:
@t-b it appears that the fit displacement issue in the single event window is not related to the use of arrow keys in the stats window. I see it when placing a cursor in the stats window on a data point from the second range: 
![image](https://github.com/AllenInstitute/MIES/assets/5693697/1043505f-75a2-4e66-9ba9-4c440dec1b76)

timjarsky commented:
@t-b when attempting to plot "stats" plots in the same plot, I get:

```
  BUG_TS: Encountered pending RTE: 385, AppendToGraph;Can't append different x-wave to category axis.
```
```
SF code:

psxStats(test, $epoch,$sel, tau, acc, stats)

with

psxStats(test, $epoch,$sel, xinterval,acc,stats)

with

psxStats(test, $epoch,$sel, amp,acc,stats)
```

I believe the x-categories are the same. Only the x-axis labels are different.

![image](https://github.com/AllenInstitute/MIES/assets/5693697/f5cd0c0a-65ea-4f8a-8da7-70fa84798c05)

timjarsky commented:
@t-b for next chat: consider adding single event alignment options to the GUI; peak and peak rate of rise (between onset and peak, may require smoothing), in addition to event-onset-based alignment.

t-b commented:
- [ ] Fix fit range gathering
- [ ] Add alignment option using peak instead of onset, add popupmenu to choose, onset is the default
- [ ] Add `riseTime` option for `psxStats`

```
rt=riseTime([lowerThreshold, upperThreshold])
psxStats(id, $epoch, $sel,$rt , acc, stats) (bearbeitet) 
```
- [ ] Fix bug with multiple psxStats with `with` and category axis
- [ ]  All points histogram, https://github.com/AllenInstitute/MIES/issues/1794

t-b commented:
I've added "Fit Result" to the legend output and added some more assertions for invalid fit ranges. This should help debug that further as I could not reproduce locally.

timjarsky commented:
> I've added "Fit Result" to the legend output and added some more assertions for invalid fit ranges. This should help debug that further as I could not reproduce locally.

@t-b I think the recipe is: place the cursor on accepted tau psxstats plot with data that has 2 or more ranges. The fit offset issue occurs if the cursor initially lands on a point in the second range. Use arrow keys to move the cursor.

t-b commented:
> @t-b I think the recipe is: place the cursor on accepted tau psxstats plot with data that has 2 or more ranges. The fit offset issue occurs if the cursor initially lands on a point in the second range. Use arrow keys to move the cursor.

Thanks for the recipe. We don't change ranges anymore when moving the cursor in the stats plot since e4c36565 (PSX: Fix mouse movement in stats plot, 2023-06-23). So at least I could not reproduce here with the latest version of this PR.

timjarsky commented:
@t-b That's a good question. Let's go with the left vertical arrow.

timjarsky commented:
@t-b When doing QC in the all-events plot with many events, it gets slow. Here are the function profiling results. Is it possible to do the heavy lifting in batches (in a thread) rather than on a per qc step?

```
Total time: 15.6851, Time in Function code: 11.3769 (72.5%)
Top function percentages:
Function MIES_EnhancedWMRoutines.ipf MIES_fWaveAverage: 86%

Annotated Top Functions:

*******************************************************************************************
Function: MIES_EnhancedWMRoutines.ipf MIES_EnhancedWMRoutines.ipf#MIES_fWaveAverage; Percent total 86%
*******************************************************************************************
[00]*         	|threadsafe Function/WAVE MIES_fWaveAverage(WAVE/Z yWaves, variable ignoreNaNs, variable averageWaveType, [variable getComponents, WAVE/WAVE prevAvgData])
[00]          	|
[00]          	|	// check the input waves, and choose an appropriate algorithm
[00]*         	|	Variable maxLength = 0
[00]          	|	Variable differentLengths= 0
[00]          	|	Variable differentXRanges= 0
[00]*         	|	Variable rawDeltaX=NaN			// 6.35: keep track of common deltaX for waveforms.
[00]          	|	Variable differentDeltax= 0	// 6.35: use interpolation if deltax's are different, even if simply reversed in sign.
[00]          	|	Variable thisXMin, thisXMax, thisDeltax
[00]          	|	Variable minXmin, maxXmax, minDeltax
[00]*         	|	Variable numXWaves=0
[00]          	|	Variable XWavesAreSame=1	// assume they are until proven differently. Irrelevant if	numXWaves!=numWaves
[00]          	|	Variable i, numWaves, gotPrevAvgData
[00]          	|
[00]*         	|	ignoreNaNs = !!ignoreNaNs
[00]          	|	getComponents = ParamIsDefault(getComponents) ? 0 : !!getComponents
[00]*         	|	Make/FREE/WAVE/N=3 result
[00]          	|
[00]*         	|	if(!WaveExists(yWaves))
[00]          	|		if(ParamIsDefault(prevAvgData))
[00]          	|			return result
[00]          	|		endif
[00]          	|		if(!(WaveExists(prevAvgData[0]) && WaveExists(prevAvgData[1])))
[00]          	|			return result
[00]          	|		endif
[00]          	|		// return the the previous avg data
[00]          	|		Duplicate/FREE prevAvgData[0], aveW
[00]          	|		WAVE sumW = prevAvgData[0]
[00]          	|		WAVE countW = prevAvgData[1]
[00]          	|		aveW = sumW / countW
[00]          	|
[00]          	|		result = {aveW, sumW, countW}
[00]*         	|		return result
[00]          	|	endif
[00]          	|
[00]*         	|	numWaves = DimSize(ywaves, ROWS)
[00]          	|	if(numWaves == 0)
[00]          	|		return result
[00]          	|	endif
[00]          	|	WAVE/WAVE waves = yWaves
[00]          	|
[00]          	|	if(!ParamIsDefault(prevAvgData))
[00]          	|		ASSERT_TS(WaveExists(prevAvgData), "Specified previous average data wave does not exist")
[00]          	|		ASSERT_TS(DimSize(prevAvgData, ROWS) == 2, "Specified previous average data wave must have exactly two rows, the sums and the counts.")
[00]          	|		gotPrevAvgData = WaveExists(prevAvgData[0]) && WaveExists(prevAvgData[1])
[00]          	|		if(gotPrevAvgData)
[00]          	|			Redimension/N=(numWaves + 1) waves
[00]          	|			waves[numWaves] = prevAvgData[0]
[00]          	|			numWaves += 1
[00]          	|		endif
[00]          	|	endif
[00]          	|
[00]*         	|	Make/D/N=(numWaves,2)/FREE xRange	// [i][0] is xMin, [i][1] is xMax
[00]          	|
[00]*         	|	for (i = 0; i < numWaves; i += 1)
[00]*         	|		Wave/Z w=waves[i]
[00]*         	|		ASSERT_TS(WaveExists(w), "A wave in the list of waves cannot be found.")
[00]          	|
[00]*         	|		Variable thisLength= numpnts(w)
[00]*         	|		thisDeltax= deltax(w)
[00]*         	|		thisXMin= leftx(w)
[00]*         	|		thisXMax= pnt2x(w, numpnts(w) - 1)
[00]*         	|		XWavesAreSame=0	// at least 1 y wave has no x wave
[00]          	|		// 6.35: point-for-point averaging requires the deltaX of all waves to be identical.
[00]*         	|		if( numtype(rawDeltaX) != 0 )
[00]*         	|			rawDeltaX = thisDeltaX	// remember first deltaX before abs() below
[00]*         	|		elseif( thisDeltax != rawDeltaX )
[00]*         	|			differentDeltax= 1	// don't do point-for-point averaging.
[00]          	|		endif
[00]*         	|		xRange[i][0]= thisXMin
[00]*         	|		xRange[i][1]= thisXMax
[00]*         	|		if( i > 0 )
[00]*         	|			if( thisLength != maxLength )
[00]*         	|				differentLengths= 1
[00]          	|			endif
[00]*         	|			if( (thisXMin != minXmin) || (thisXMax != maxXmax) )
[00]*         	|				differentXRanges= 1	// this also includes the case where identical ranges but one or more is swapped
[00]          	|			endif
[00]*         	|			if( i == 1 )
[00]          	|				// handle case where first wave's x range is swapped.
[00]*         	|				if( minXmin > maxXmax )	// swapped X range (X values decrease with increasing point number)
[00]          	|					Variable tmp= minXmin
[00]          	|					minXmin= maxXmax
[00]*         	|					maxXmax= tmp
[00]          	|				endif
[00]          	|			endif
[00]*         	|			if( thisXMin > thisXMax )	// swapped X range (X values decrease with increasing point number)
[00]          	|				tmp= thisXMin
[00]          	|				thisXMin= thisXMax
[00]*         	|				thisXMax= tmp
[00]          	|			endif
[00]          	|			// accumulate x ranges
[00]*         	|			minXmin= min(minXmin, thisXMin)
[00]*         	|			maxXmax= max(maxXmax, thisXMax)
[00]          	|			// find smallest deltax
[00]*         	|			thisDeltax= abs(thisDeltax)
[00]*         	|			if( thisDeltax > 0 && (thisDeltax < minDeltax) )
[00]*         	|				minDeltax= thisDeltax
[00]*         	|			endif
[00]          	|		else
[00]*         	|			minXmin= thisXMin
[00]          	|			maxXmax= thisXMax
[00]          	|			minDeltax= abs(thisDeltax)
[00]*         	|			if( minDeltax == 0 )
[00]*         	|				thisDeltax= inf
[00]          	|			endif
[00]          	|		endif
[00]*         	|		maxLength = max(maxLength, thisLength)
[00]*         	|	endfor
[00]          	|
[00]          	|	Variable doPointForPoint
[00]*         	|	doPointForPoint = (!differentXRanges && !differentLengths && !differentDeltax) || numtype(minDeltaX) != 0 || minDeltaX == 0
[00]          	|
[00]*         	|	if(doPointForPoint)
[00]          	|		Concatenate/FREE {waves}, fullWave
[00]          	|
[00]          	|		if(!getComponents)
[00]          	|			if(ignoreNaNs)
[00]          	|				if(averageWaveType == IGOR_TYPE_32BIT_FLOAT)
[00]          	|					MatrixOP/FREE AveW = fp32(sumRows(replaceNaNs(fp64(fullWave), 0)) / sumRows(equal(numtype(fullWave), 0)))
[00]          	|				elseif(averageWaveType == IGOR_TYPE_64BIT_FLOAT)
[00]          	|					MatrixOP/FREE AveW = sumRows(replaceNaNs(fp64(fullWave), 0)) / sumRows(equal(numtype(fullWave), 0))
[00]          	|				else
[00]          	|					ASSERT_TS(0, "Not supported")
[00]          	|				endif
[00]          	|			else
[00]          	|				Make/FREE/N=(maxLength)/Y=(averageWaveType) AveW
[00]          	|				if(averageWaveType == IGOR_TYPE_32BIT_FLOAT)
[00]          	|					MatrixOP/FREE AveW = fp32(sumRows(fp64(fullWave)) / numWaves)
[00]          	|				elseif(averageWaveType == IGOR_TYPE_64BIT_FLOAT)
[00]          	|					MatrixOP/FREE AveW = sumRows(fp64(fullWave)) / numWaves)
[00]          	|				else
[00]          	|					ASSERT_TS(0, "Not supported")
[00]          	|				endif
[00]          	|			endif
[00]          	|		else
[00]          	|			if(ignoreNaNs)
[00]          	|				if(averageWaveType == IGOR_TYPE_32BIT_FLOAT)
[00]          	|					MatrixOP/FREE SumW = sumRows(replaceNaNs(fp64(fullWave), 0)
[00]          	|					MatrixOP/FREE CntW = sumRows(equal(numtype(fullWave), 0)))
[00]          	|					if(gotPrevAvgData)
[00]          	|						WAVE prevCnt = prevAvgData[1]
[00]          	|						MatrixOp/FREE CntW = CntW + prevCnt - 1
[00]          	|					endif
[00]          	|					MatrixOP/FREE AveW = fp32(SumW / CntW)
[00]          	|				elseif(averageWaveType == IGOR_TYPE_64BIT_FLOAT)
[00]          	|					MatrixOP/FREE SumW = sumRows(replaceNaNs(fp64(fullWave), 0))
[00]          	|					MatrixOP/FREE CntW = sumRows(equal(numtype(fullWave), 0))
[00]          	|					if(gotPrevAvgData)
[00]          	|						WAVE prevCnt = prevAvgData[1]
[00]          	|						MatrixOp/FREE CntW = CntW + prevCnt - 1
[00]          	|					endif
[00]          	|					MatrixOP/FREE AveW = SumW / CntW
[00]          	|				else
[00]          	|					ASSERT_TS(0, "Not supported")
[00]          	|				endif
[00]          	|			else
[00]          	|				Make/FREE/N=(maxLength)/Y=(averageWaveType) AveW
[00]          	|				if(averageWaveType == IGOR_TYPE_32BIT_FLOAT)
[00]          	|					MatrixOP/FREE SumW = sumRows(fp64(fullWave))
[00]          	|					if(gotPrevAvgData)
[00]          	|						WAVE prevCnt = prevAvgData[1]
[00]          	|						MatrixOP/FREE CntW = prevCnt + numWaves - 1
[00]          	|						MatrixOP/FREE AveW = fp32(SumW / CntW)
[00]          	|					else
[00]          	|						MatrixOP/FREE AveW = fp32(SumW / numWaves)
[00]          	|						Duplicate/FREE SumW, CntW
[00]          	|						FastOp CntW = (numWaves)
[00]          	|					endif
[00]          	|				elseif(averageWaveType == IGOR_TYPE_64BIT_FLOAT)
[00]          	|
[00]          	|					MatrixOP/FREE SumW = sumRows(fp64(fullWave))
[00]          	|					if(gotPrevAvgData)
[00]          	|						WAVE prevCnt = prevAvgData[1]
[00]          	|						MatrixOP/FREE CntW = prevCnt + numWaves - 1
[00]          	|						MatrixOP/FREE AveW = SumW / CntW
[00]          	|					else
[00]          	|						MatrixOP/FREE AveW = SumW / numWaves
[00]          	|						Duplicate/FREE SumW, CntW
[00]          	|						FastOp CntW = (numWaves)
[00]          	|					endif
[00]          	|				else
[00]          	|					ASSERT_TS(0, "Not supported")
[00]          	|				endif
[00]          	|			endif
[00]          	|
[00]          	|			CopyScales/P waves[0], SumW, CntW
[00]          	|		endif
[00]          	|
[00]          	|		CopyScales/P waves[0], AveW
[00]          	|
[00]          	|		result = {AveW, SumW, CntW}
[00]*         	|		return result
[00]          	|	else
[00]          	|		// can't do point-for-point because of different point range or scaling
[00]          	|		Variable firstAvePoint,lastAvePoint,point,xVal,yVal
[00]          	|
[00]*         	|		Variable newLength= 1 + round(abs(maxXmax - minXmin) / minDeltaX)
[00]*         	|		maxLength= min(maxLength*4,newLength)	// avoid the case where one very small deltaX in an XY pair causes a huge wave to be created.
[00]          	|
[00]*         	|		Make/FREE/N=(maxLength)/Y=(averageWaveType) AveW, TempNWave, TempYWave
[00]*         	|		Wave w = waves[0]
[00]*         	|		CopyScales w, AveW // just to get the data and x units
[00]*         	|		SetScale/I x, minXmin, maxXmax, AveW // set X scaling to all-encompassing range
[00]          	|
[00]*         	|		for (i = 0; i < numWaves; i += 1)
[00]*         	|			thisXMin= xRange[i][0]
[00]*         	|			thisXMax= xRange[i][1]
[00]*         	|			if( thisXMin > thisXMax )	// swapped X range (X values decrease with increasing point number)
[00]          	|				tmp= thisXMin
[00]          	|				thisXMin= thisXMax
[00]*         	|				thisXMax= tmp
[00]          	|			endif
[00]*         	|			firstAvePoint= ceil(x2pntWithFrac(AveW,thisXMin))	// truncate the partial point numbers...
[00]*         	|			lastAvePoint= floor(x2pntWithFrac(AveW,thisXMax))	// ... by indenting slightly
[00]*         	|			WAVE wy = waves[i]
[28]***       	|			MultiThread TempYWave[firstAvePoint, lastAvePoint] = wy(limit(pnt2x(AveW, p), thisXMin, thisXMax))
[00]          	|
[00]*         	|			if(ignoreNaNs)
[29]***       	|				MultiThread AveW[firstAvePoint, lastAvePoint]      += !IsNaN(TempYWave[p]) ? TempYWave[p] : 0
[28]***       	|				MultiThread TempNWave[firstAvePoint, lastAvePoint] += !IsNaN(TempYWave[p])
[00]          	|			else
[00]          	|				MultiThread AveW[firstAvePoint, lastAvePoint] += TempYWave[p]
[00]*         	|				MultiThread TempNWave[firstAvePoint, lastAvePoint] += 1
[00]          	|			endif
[00]*         	|		endfor
[00]          	|
[00]          	|		if(getComponents)
[00]          	|			Duplicate/FREE AveW, SumW
[00]          	|			CopyScales/P AveW, TempNWave
[00]          	|		endif
[00]          	|
[00]*         	|		if(gotPrevAvgData)
[00]          	|			WAVE prevCnt = prevAvgData[1]
[00]          	|			Make/FREE/N=(maxLength)/Y=(averageWaveType) prevCntXAdapt
[00]          	|			MultiThread prevCntXAdapt[firstAvePoint, lastAvePoint] = prevCnt(limit(pnt2x(AveW, p), thisXMin, thisXMax))
[00]*         	|			MultiThread TempNWave[firstAvePoint, lastAvePoint] += prevCntXAdapt - 1
[00]          	|		endif
[00]          	|
[00]          	|		//  points with no values added are set to NaN here:
[00]*         	|		MultiThread AveW= (TempNWave[p] == 0) ? NaN : AveW[p] / TempNWave[p]
[00]          	|
[00]*         	|		result = {AveW, SumW, TempNWave}
[00]          	|		return result
[00]          	|	endif
[00]          	|End
```

timjarsky commented:
@t-b using the marquee tool to select a batch of traces is another viable alternative

t-b commented:
@timjarsky The new rise time property should use event or fit QC?

timjarsky commented:
> @timjarsky The new rise time property should use event or fit QC?

@t-b good question. Event QC is correct.

t-b commented:
@timjarsky Ready for another try!

t-b commented:
@timjarsky Ready for another try!

t-b commented:
Let's also the enjoy the finished todo list as long as it holds 🎇

timjarsky commented:
@t-b PSX stats event plots don't indicate what is being plotted:

<img width="611" alt="image" src="https://github.com/AllenInstitute/MIES/assets/5693697/77f4e6f8-7981-47a8-b0eb-39df00e2b297">

Could the ```prop:     amp/xpos/xinterval/tau/estate/fstate/fitresult/risetime``` be added to the axis label?

timjarsky commented:
@t-b with Jessica's NWB 2023_06_27_161103-compressed.nwb (uploaded to your FTP) I get:

```
  !!! Threadsafe assertion FAILED !!!
  Message: "Could not find upper level crossing"
  Please provide the following information if you contact the MIES developers:
  ################################
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Stacktrace:
  PSX_CalculateRiseTimeImpl(...)#L1349 [MIES_SweepFormula_PSX.ipf]
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Time: 2023-07-28T11:50:16-07:00
  Experiment: Untitled ()
  Igor Pro version: 9.0.2.4 (39951)
  ################################
  !!! Threadsafe assertion FAILED !!!
  Message: "Could not find lower level crossing"
  Please provide the following information if you contact the MIES developers:
  ################################
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Stacktrace:
  PSX_CalculateRiseTimeImpl(...)#L1345 [MIES_SweepFormula_PSX.ipf]
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Time: 2023-07-28T11:50:16-07:00
  Experiment: Untitled ()
  Igor Pro version: 9.0.2.4 (39951)
  ################################

```

with

```
sel=select(channels(AD9),[31,32,35,37,39],all) 
kernel = psxKernel(E0,$sel, 0.45, 1, (-6))

psx(quack,$kernel, 0.046, 550, 0)

and

psxprep(psx(quack,$kernel, 0.046, 550, 0),2.5)
```
On the first Display, but not after.

t-b commented:
> @t-b PSX stats event plots don't indicate what is being plotted

What was the exact `psxStats` formula? Here I always get a description, on either the y or the x axis.

t-b commented:
> On the first Display, but not after.

The results are cached, thererfore we do the computation only once.

t-b commented:
- [ ] Marquee selection, outside of graph ?!, does not work sometimes

t-b commented:
@timjarsky Ready for another try. Only documentation and some tests are missing. The new stats post processing is called "nonfinite".

t-b commented:
@timjarsky Ready

timjarsky commented:
@t-b this is looking great. I found only one potential issue:

```psxStats(quack, E0,$sel,  rise, acc, nonfinite)```

makes this plot

<img width="905" alt="image" src="https://github.com/AllenInstitute/MIES/assets/5693697/47d88f64-cd02-4414-8fea-d8ef9243b862">


But placing the cursor on the plot doesn't select the event

In this example, I expect it to select event 160, but it stays on event zero.

<img width="338" alt="image" src="https://github.com/AllenInstitute/MIES/assets/5693697/c28c5e58-d86e-4242-8d03-325d907b438e">

t-b commented:
One can either say I missed mouse cursor support or that it was not ordered ;) In any case I'll fix/add it.

t-b commented:
@timjarsky Should be ready again.
